src/lib/component/autocomplete/index.js

```diff
@@ -0,0 +1,193 @@
+import AutocompleteModel from './autocompleteModel'
+import debounce300 from '../debounce300'
+import delegate from 'delegate'
+
+export default class Autocomplete {
+  #inputElement
+  #onSelect
+  #onPreview
+  #model
+  #resultsElement
+
+  constructor(inputElement, onSearch, onSelect, onPreview) {
+    this.#inputElement = inputElement
+    this.#onSelect = onSelect
+    this.#onPreview = onPreview
+
+    this.#model = new AutocompleteModel(
+      (term) => onSearch(term, (results) => (this.#model.items = results)),
+      (items) => this.#renderItem(items),
+      (index) => this.#highlight(index)
+    )
+
+    this.#resultsElement = document.createElement('ul')
+    this.#resultsElement.setAttribute('popover', 'auto')
+    this.#resultsElement.classList.add('textae-editor__dialog__autocomplete')
+    inputElement.parentElement.appendChild(this.#resultsElement)
+
+    const handleInput = debounce300((term) => {
+      this.#model.term = term
+
+      if (term.length < 3) {
+        this.#resultsElement.hidePopover()
+      }
+    })
+
+    this.#inputElement.addEventListener('input', (event) =>
+      handleInput(event.target.value)
+    )
+    this.#inputElement.addEventListener('keydown', (event) =>
+      this.#handleKeydown(event)
+    )
+
+    // Hide popover when input is out of focus.
+    this.#inputElement.addEventListener('blur', () =>
+      this.#resultsElement.hidePopover()
+    )
+
+    this.#setEventHandlerToResults()
+  }
+
+  #setEventHandlerToResults() {
+    delegate(this.#resultsElement, 'li', 'mousedown', (e) => {
+      this.#onSelect(
+        e.delegateTarget.dataset.id,
+        e.delegateTarget.dataset.label
+      )
```

Comment:
JavaScriptでは引数でもデストラクチャリングが使えます。
次のように引数eのプロパティを引数で受け取れます。

```suggestion
    delegate(this.#resultsElement, 'li', 'mousedown', ({delegateTarget}) => {
      this.#onSelect(
        delegateTarget.dataset.id,
        delegateTarget.dataset.label
      )
```

今回のように delegateTarget しか参照しない場合に、少し短くかけます。

Reply:
修正しました。
https://github.com/pubannotation/textae/pull/142/commits/de4fa95544855e04d4bea448f8f0c6c334036da6
https://github.com/pubannotation/textae/pull/142/commits/03c45761019fd6281c961723fa43627d57239cd7